central/complianceoperator/v2/scanconfigurations/service/download.go

```diff
@@ -0,0 +1,138 @@
+		return
+	}
+
+	ctx := r.Context()
+	slimUser := authn.UserFromContext(ctx)
+	if slimUser == nil {
+		httputil.WriteGRPCStyleError(w, codes.PermissionDenied, errors.New("Could not determine user identity from provided context"))
+		return
+	}
+
+	snapshot, found, err := h.snapshotDataStore.GetSnapshot(ctx, id)
+	if err != nil {
+		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Error retrieving snapshot with the given id %s", id))
+		return
+	}
+	if !found {
+		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Unable to find the snapshot with the given id %s", id))
+		return
+	}
+
+	if slimUser.GetId() != snapshot.GetUser().GetId() {
+		httputil.WriteGRPCStyleError(w, codes.PermissionDenied, errors.Errorf("Report %s cannot be downloaded by the user %s", id, slimUser.GetId()))
+		return
+	}
+
+	status := snapshot.GetReportStatus()
+	if status.GetReportNotificationMethod() != storage.ComplianceOperatorReportStatus_DOWNLOAD {
+		httputil.WriteGRPCStyleError(w, codes.InvalidArgument, errors.Errorf("Report %s did not generated a downloaded report", id))
+		return
+	}
+
+	switch status.GetRunState() {
+	case storage.ComplianceOperatorReportStatus_FAILURE:
+		httputil.WriteGRPCStyleError(w, codes.FailedPrecondition, errors.Errorf("Report %s failed", id))
+		return
+	case storage.ComplianceOperatorReportStatus_WAITING, storage.ComplianceOperatorReportStatus_PREPARING:
+		httputil.WriteGRPCStyleError(w, codes.Unavailable, errors.Errorf("Report %s is not ready for download", id))
+		return
+	}
+
+	buf := bytes.NewBuffer(nil)
+	ctx = sac.WithGlobalAccessScopeChecker(ctx,
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+			sac.ResourceScopeKeys(resources.Administration),
+		),
+	)
+
+	_, exists, err := h.blobStore.Get(ctx, common.GetComplianceReportBlobPath(snapshot.GetScanConfigurationId(), id), buf)
+	if err != nil {
+		httputil.WriteGRPCStyleError(w, codes.Internal, errors.New("Failed to fetch report data"))
+		return
+	}
+	if !exists {
+		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Report %s is not available to download", id))
+		return
+	}
+
+	// Tell the browser this is a download.
+	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename="report-%s.zip"`, zip.GetSafeFilename(id)))
+	w.Header().Set("Content-Type", "application/zip")
+	w.Header().Set("Content-Length", fmt.Sprint(buf.Len()))
+	_, err = w.Write(buf.Bytes())
+
+	writeSnapshotCtx := sac.WithGlobalAccessScopeChecker(ctx,
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_WRITE_ACCESS),
+			sac.ResourceScopeKeys(resources.Compliance)),
+	)
+	if err == nil && status.GetRunState() == storage.ComplianceOperatorReportStatus_GENERATED {
+		snapshot.GetReportStatus().RunState = storage.ComplianceOperatorReportStatus_DELIVERED
+		err = h.snapshotDataStore.UpsertSnapshot(writeSnapshotCtx, snapshot)
+		if err != nil {
+			log.Error("Error setting report state to DELIVERED")
+		}
+	}
```

Comment:
The error from `w.Write`  is unhandled at the end of this block. It may have been already overwritten by the `UpsertSnapshot` error. I think we should handle it immediately in line 113

Reply:
Done